Pull in Upstream Updates

mcdonnnj opened the pull request:
# <!--- Provide a general summary of your changes in the Title above -->

## 🗣 Description
This PR is to pull in changes from cisagov/skeleton-generic#41.
After a long period of sorting out updating to include necessary [cisagov/skeleton-packer-cool](https://github.com/cisagov/skeleton-packer-cool) elements, this now also includes cisagov/skeleton-generic#42 and cisagov/skeleton-generic#43.

The project has been updated to use [cisagov/ami-build-iam-user-tf-module](https://github.com/cisagov/ami-build-iam-user-tf-module) instead of [cisagov/molecule-packer-ci-iam-user-tf-module](https://github.com/cisagov/molecule-packer-ci-iam-user-tf-module) for creating the `test-skeleton-packer` user. This has required a transition to building in the COOL, and breaks backwards compatibility with this project and non-COOL Packer projects we have developed.
<!--- Describe your changes in detail -->

## 💭 Motivation and Context
Updates and bugfixes are good.
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## 🧪 Testing
Automated testing passed.

I was able to successfully create the `test-skeleton-packer` user in the COOL and successfully build an AMI using the local environment setup. I verified it was found in the list of AMIs in the `Images (Staging)` account on AWS.
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## 📷 Screenshots (if appropriate)

## 🚥 Types of Changes

<!--- What types of changes does your code introduce? -->
<!--- Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (causes existing functionality to change)

## ✅ Checklist

<!--- Go over all the following points, and put an `x` in all the
boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask.
We're here to help! -->

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

mcdonnnj committed:
Ran `pre-commit autoupdate` to get the latest version of the
`pre-commit-terraform` hook since the PR I submitted,
https://github.com/antonbabenko/pre-commit-terraform/pull/100,
was approved. This will fix issues with `skeleton-tf-module` related to multiple
directories with Terraform code.

mcdonnnj committed:
Explicitly install pre-commit hooks as its own step so it's clearer when the
failure is with setup rather than hooks running.

mcdonnnj committed:
Rename action to accurately reflect that we are only linting, not building.
Adjust description for pre-commit hook setup to better convey what is being
done.

mcdonnnj committed:
Sort .gitignore entries.

mcdonnnj committed:
Merge pull request #41 from cisagov/bugfix/update_pre-commit_for_pre-commit-terraform_hook

Update pre-commit Hooks to Include Bugfix

mcdonnnj committed:
Merge branch 'develop' of github.com:cisagov/skeleton-generic into improvements/pull_upstream_updates

jsf9k reviewed (dismissed):
Huzzah!

mcdonnnj committed:
Add Terraform related installation and initialization from skeleton-tf-module
to the 'lint' Action to provide necessary `pre-commit` elements.

felddy reviewed (dismissed):
Boosh!  🌩

mcdonnnj committed:
Rename the 'terraform' directory to 'terraform-test-user' to line up with the
'skeleton-packer-cool' directory structure.

mcdonnnj committed:
Add '.terraform' directories to the '.gitignore' file.

hillaryj committed:
Update CODEOWNERS

mcdonnnj committed:
Add the 'terraform-post-packer' direction from skeleton-packer-cool.

mcdonnnj committed:
Update the 'terraform-test-user' Terraform code to mirror skeleton-packer-cool
methods.

mcdonnnj committed:
Fix references to skeleton-packer-cool in the 'terraform-post-packer' directory.

mcdonnnj committed:
Update the prerelease workflow to mirror skeleton-packer-cool.

mcdonnnj committed:
Update the release workflow to mirror skeleton-packer-cool.

hillaryj committed:
Merge pull request #42 from cisagov/hillaryj-owner

Update CODEOWNERS

felddy committed:
Allow events from apb to rebuild this repository weekly.

felddy committed:
Merge pull request #43 from cisagov/improvement/apb-events

Allow events from apb to rebuild this repository weekly.

mcdonnnj committed:
Remove the terraform-post-packer directory because it relies on the COOL account
structure to function. The base `skeleton-packer` should still play nice with
non-COOL projects for the time being.
Update the `terraform-test-user` Terraform:
-Move the provider blocks from `main.tf` into their own file (`providers.tf`).
-Move the `caller_user_name` local definition into `providers.tf`.
-Move the `aws_caller_identity` block into its own file (`caller_identity.tf`).
-Fix the `caller_user_name` local variable to correctly indicate the username
 of the caller. Previously the split would return the name of the assumed role
 instead of the username of the caller.

mcdonnnj committed:
Merge branch 'develop' of github.com:cisagov/skeleton-generic into improvements/pull_upstream_updates

mcdonnnj committed:
Update theGitHub Actions workflow to use actions/setup-go@v2 instead of v2-beta.
Update the version of Go installed to the latest stable version.
Update the installation of terraform-docs so that it correctl installs.
Please see https://github.com/cisagov/skeleton-tf-module/pull/19 for more
information about this change.

mcdonnnj committed:
Added jsf9k's fix for Terraform code that uses `remote_state` not playing well
with the current `terraform validate` implementation:
I am commenting out this hook because terraform validate cannot currently run
without accessing the remote state because of the way the providers are defined
in providers.tf. This is something that may be remedied in the future. For more
information, check out these two GitHub issues:
* https://github.com/hashicorp/terraform/issues/15895
* https://github.com/hashicorp/terraform/issues/15811

mcdonnnj committed:
Re-add `terraform-post-packer` since the priority is COOL environment needs.
Update the README to correctly reflect building this project as it is now.
Add tweaks from the `packer.json` configuration in `skeleton-packer-cool`.

mcdonnnj committed:
Fix some dangling `skeleton-packer-cool` references.

mcdonnnj committed:
Adjust the line break for pre-commit in the GitHub Actions workflow.

mcdonnnj committed:
Remove line break character for pre-commit in GitHub Actions workflow.

dav3r commented on .github/workflows/build.yml:
```diff
@@ -40,14 +46,63 @@ jobs:
+        # * https://github.com/hashicorp/terraform/issues/15811
         run: pre-commit run --all-files
-
+                            --config=.github-actions-pre-commit-config.yaml
```
Let's standardize this syntax the way @jsf9k did it [here](https://github.com/cisagov/cool-assessment-terraform/blob/develop/.github/workflows/build.yml#L104-L106).

mcdonnnj replied:
Ah, I used [this syntax](https://github.com/cisagov/ansible-role-nessus/blob/c42a575a7683b1f036352aae207d8d3983897c64/.github/workflows/build.yml#L104) as a reference (and had to break the line). Resolved in 022fdf0.

dav3r replied:
It wasn't a problem in `ansible-role-nessus` because we have the [line length check disabled](https://github.com/cisagov/ansible-role-nessus/blob/c42a575a7683b1f036352aae207d8d3983897c64/.yamllint#L28) in our ansible repos.  In general, we should use the syntax I mentioned above.

dav3r commented on README.md:
```diff
@@ -4,21 +4,35 @@
 This is a generic skeleton project that can be used to quickly get a
 new [cisagov](https://github.com/cisagov) GitHub
-[Packer](https://packer.io) project started.  This skeleton project
+[Packer](https://packer.io) project started. This skeleton project
```
Oh boy, here we go......  🤯

jsf9k replied:
:rage:

mcdonnnj replied:
![betty_white_dabbing](https://user-images.githubusercontent.com/50747025/81223912-d6de4e80-8fb4-11ea-8a9a-4138280f3d04.gif)

jsf9k commented on .github-actions-pre-commit-config.yaml:
```diff
@@ -0,0 +1,113 @@
+    rev: v1.29.0
+    hooks:
+      - id: terraform_fmt
+      # I am commenting out this hook because terraform validate
```
I was thinking about the way we are doing this just this morning.  With separate files, the two will eventually fall out of sync with each other.  I think what makes more sense is to do the following in the GitHub Actions workflow:
1. Run `sed` or something similar against the default `.pre-commit-config.yaml` file to comment out the `terraform_validate` line and generate a new file `/tmp/.github-actions-pre-commit-config.yaml`.
2. Run `pre-commit` with the `/tmp/.github-actions-pre-commit-config.yaml` configuration.

dav3r replied:
Ah, that is a very good idea!

mcdonnnj replied:
Fantastic idea! Implemented in 022fdf0 (and fixed in 0724b04).

jsf9k replied:
Excellent work @mcdonnnj!

dav3r commented on README.md:
```diff
@@ -4,21 +4,35 @@
-code will create the user with the appropriate name and permissions.  This only
-needs to be run once per project, per AWS account.  This user will also be used by
-GitHub Actions.
+This project requires a build user to exist in AWS. The accompanying Terraform
+code will create the user with the appropriate name and permissions. This only
+needs to be run once per project, per AWS account. This user will also be used
+by GitHub Actions.
+
+Before the build user can be created, the following profile must exist in
+your AWS credentials file:
+
+* `cool-terraform-backend`
+
+The easiest way to set up that profile is to use our
+[`aws-profile-sync`](https://github.com/cisagov/aws-profile-sync) utility.
+Follow the usage instructions in that repository before continuing with the
+next steps. Note that you will need to know where your team stores their
+remote profile data in order to use
+[`aws-profile-sync`](https://github.com/cisagov/aws-profile-sync).
+
+To create the build user, follow these instructions:
```
@jsf9k Do you have the same concerns here calling this a "build user" as [you did in our ansible roles](https://github.com/cisagov/ansible-role-nessus/pull/8#pullrequestreview-397446385)?  

In these packer repos, I think "build user" is appropriate.  If we are all in agreement, then I think we should change the `terraform-test-user` directory to `terraform-build-user` and update the contents accordingly (`test` -> `build`).

jsf9k replied:
I'm fine with the packer user being called a "build user".

I agree with the rest of what you said as well.

mcdonnnj replied:
Should the created username also be changed? Right now it is `test-skeleton-packer`, and if we change the directory to `terraform-build-user` that seems inconsistent.

dav3r replied:
Yes, I'm good with renaming the build user to `build-skeleton-packer`.

mcdonnnj replied:
These changes have been implemented in ec21791.

dav3r replied:
You missed a spot with your renaming:  https://github.com/cisagov/skeleton-packer/blob/improvements/pull_upstream_updates/terraform-build-user/backend.tf#L8

mcdonnnj replied:
That was only changed from `skeleton-packer-cool` to `skeleton-packer` (https://github.com/cisagov/skeleton-packer-cool/blob/590f4fc3d5c2a0963d718f824b4b07185fe5c25a/terraform-test-user/terraform.tf#L8). Both `terraform-build-user` and `terraform-post-packer` are using `skeleton-packer` as the base for their state keys.

dav3r replied:
The `skeleton-packer` part is fine- I was talking about changing `terraform-test-user.tfstate` to `terraform-build-user.tfstate`.

mcdonnnj replied:
Ahh, my misunderstanding. Fixed in 8432b1a.

dav3r replied:
Excellent!

dav3r commented on terraform-post-packer/main.tf:
```diff
@@ -0,0 +1,79 @@
+# Default AWS provider (EC2AMICreate role in the Images account)
+provider "aws" {
+  region  = "us-east-1"
+  profile = "cool-images-ec2amicreate"
+}
+
+# AWS provider for the Master account (OrganizationsReadOnly role)
+provider "aws" {
+  region  = "us-east-1"
+  profile = "cool-master-organizationsreadonly"
+  alias   = "master"
+}
+
+# Use aws_caller_identity with the default provider (Images account)
+# so we can provide the Images account ID below
+data "aws_caller_identity" "images" {
+}
```
Can you please split the providers and caller identity out to `providers.tf` and `caller_identity.tf` so that we are consistent with the code in the `terraform-test-user` directory?

mcdonnnj replied:
Whoops, forgot to do that when I re-added the `terraform-post-packer` sauce. Resolved in be9e283.

dav3r replied:
Looks great! 👍

dav3r commented on terraform-test-user/terraform.tf:
```diff
@@ -0,0 +1,10 @@
+terraform {
+  backend "s3" {
+    encrypt        = true
+    bucket         = "cisa-cool-terraform-state"
+    dynamodb_table = "terraform-state-lock"
+    profile        = "cool-terraform-backend"
+    region         = "us-east-1"
+    key            = "skeleton-packer/terraform-test-user.tfstate"
+  }
+}
```
This file should be renamed `backend.tf` to match how we do it elsewhere.

mcdonnnj replied:
Good catch, resolved in 8ed3842.

dav3r commented on terraform-post-packer/terraform.tf:
```diff
@@ -0,0 +1,10 @@
+terraform {
+  backend "s3" {
+    encrypt        = true
+    bucket         = "cisa-cool-terraform-state"
+    dynamodb_table = "terraform-state-lock"
+    profile        = "cool-terraform-backend"
+    region         = "us-east-1"
+    key            = "skeleton-packer/terraform-post-packer.tfstate"
+  }
+}
```
This file should be renamed `backend.tf` to match how we do it elsewhere.

mcdonnnj replied:
Good catch, resolved in 8ed3842.

mcdonnnj committed:
Rename `terraform.tf` files to `backend.tf` for consistency with other projects
per feedback.

mcdonnnj committed:
Break out providers and caller_identify from `main.tf` to their own files in the
`terraform-post-packer` directory.

mcdonnnj committed:
Modify the GitHub Actions workflow to generate a temporary pre-commit
configuration file with the terraform_validate hook removed.
Switch the GitHub Actions run of pre-commit to use the generated file instead.

mcdonnnj committed:
Correct the extension for the pre-commit configuration referenced in the
GitHub Actions workflow from .yml to .yaml

mcdonnnj committed:
Per discussion and feedback the terraform-test-user directory has been renamed
to terraform-build-user. Additionally all references to `test-skeleton-packer`
have been changed to `build-skeleton-packer`. A `terraform destroy` and
`terraform apply` have been run to correct these resources in AWS.

mcdonnnj committed:
Fix the terraform-build-user state key to reflect the transition from test-user
to build-user.

dav3r commented on README.md:
````diff
@@ -4,21 +4,35 @@
 
 ```console
-cd terraform
+cd terraform-test-user
````
This line needs to change to `cd terraform-build-user`.

mcdonnnj replied:
Ok I _think_ (_**hope**_) that's all of them now as of 022e5d8.

dav3r reviewed (approved):
Outstanding work on all of this @mcdonnnj - thanks for shepherding this PR through a long and arduous process!  🎉 🥇

mcdonnnj committed:
Fix a dangling test-user reference in the README. With this I think all
references to test-user have appropriately been updated (famous last words).

jsf9k reviewed (approved):
Excellent work @mcdonnnj!  Glad to finally get this one in the can.

mcdonnnj commented:
I have removed the existing Secrets for this repository and have used the `terraform-to-secrets` tool from [cisagov/development-guide](https://github.com/cisagov/development-guide) to add in updated and new Secrets now that this PR has been approved.
![image](https://user-images.githubusercontent.com/50747025/81229251-7e5f7f00-8fbd-11ea-9c9a-4de5753612f6.png)

felddy reviewed (approved):
Looks good.  Check that "Breaking change" checkbox in the description.

mcdonnnj commented:
> 
> 
> Looks good. Check that "Breaking change" checkbox in the description.

Thanks for the catch on that. Updated the `Types of Changes` and `Checklist` in the PR description to correctly reflect all the changes that have been made.
